pkg/registry/migrate.go

```diff
@@ -0,0 +1,150 @@
+package registry
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	clusterv1beta1 "github.com/replicatedhq/embedded-cluster-kinds/apis/v1beta1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const RegistryDataMigrationCompleteSecretName = "registry-data-migration-complete"
+const registryDataMigrationJobName = "registry-data-migration"
+
+const RegistryMigrationStatusConditionType = "RegistryMigrationStatus"
+const RegistryMigrationServiceAccountName = "registry-data-migration-serviceaccount"
+
+// MigrateRegistryData should be called when transitioning from non-HA to HA airgapped installations
+// this function creates a job that will scale down the registry deployment then upload the data to s3
+// before finally creating a 'migration is complete' secret in the registry namespace
+// if this secret is present, the function will return without reattempting the migration
+func MigrateRegistryData(ctx context.Context, in *clusterv1beta1.Installation, cli client.Client) error {
+	hasMigrated, err := HasRegistryMigrated(ctx, cli)
+	if err != nil {
+		return fmt.Errorf("check if registry has migrated before running migration: %w", err)
+	}
+	if hasMigrated {
+		in.Status.SetCondition(metav1.Condition{
+			Type:               RegistryMigrationStatusConditionType,
+			Status:             metav1.ConditionTrue,
+			Reason:             "MigrationJobCompleted",
+			ObservedGeneration: in.Generation,
+		})
+		return nil
+	}
+
+	// check if the migration is already in progress
+	// if it is, return without reattempting the migration
+	migrationJob := batchv1.Job{}
+	err = cli.Get(ctx, client.ObjectKey{Namespace: registryNamespace, Name: registryDataMigrationJobName}, &migrationJob)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("get migration job: %w", err)
+		}
+	} else {
+		if migrationJob.Status.Active > 0 {
+			return nil
+		}
+		if migrationJob.Status.Failed > 0 {
+			in.Status.SetCondition(metav1.Condition{
+				Type:               RegistryMigrationStatusConditionType,
+				Status:             metav1.ConditionFalse,
+				Reason:             "MigrationJobFailed",
+				ObservedGeneration: in.Generation,
+			})
+			return fmt.Errorf("registry migration job failed")
+		}
+		// TODO: handle other conditions
+		return nil
+	}
+
+	// create the migration job
+	migrationJob = batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
```

Comment:
add controller reference and apply common labels

https://github.com/replicatedhq/embedded-cluster-operator/blob/main/pkg/registry/seaweedfs.go#L67-L72